### PR TITLE
Fix config mapping when using enums

### PR DESCRIPTION
**Pull request by dpeng817**
## Summary
When a job is using config mapping, there are two validation steps that happen during execution. Once, for the outer config, and once for the mapped config. 

We use this pattern to implement default config on jobs. This does not fully adhere to the expected pattern though, because the config mapping might return already validated and post processed values, which the config system doesn't know how to pre-process.

This is particularly a problem for enum use cases, where pre-processing, we ingest enum config as strings, and convert them post process into enum values. The double-validation step ends up breaking the config system.

This introduces a toggle onto config mapping that will validate outer config, apply defaults, but not resolve to post-processing config types.

## Test Plan
Added a battery of tests for both config mapping, and default config on jobs.

## Checklist
- [x ] I have added tests to cover my changes.

### CODE REVIEW

File: python_modules/dagster/dagster/core/definitions/config.py [hunk 1]
```diff
@@ -18,6 +18,7 @@ class ConfigMapping(
         [
             ("config_fn", Callable[[Any], Any]),
             ("config_schema", IDefinitionConfigSchema),
+            ("enable_pre_processing", bool),
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/definitions/config.py:21]: don't love this name, maybe "receive_processed_config_values"

definitely deserves a clarifying comment

File: python_modules/dagster/dagster/core/system_config/objects.py [hunk 2]
```diff
@@ -152,18 +167,28 @@ def build(
...
             if not outer_evr.success:
                 raise DagsterInvalidConfigError(
                     f"Error in config mapping for pipeline {pipeline_def.name} mode {mode}",
                     outer_evr.errors,
                     run_config,
                 )
+            outer_config = resolve_defaults(
+                cast("ConfigType", run_config_schema.config_mapping.config_schema.config_type),
+                outer_evr.value,
+            ).value
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/system_config/objects.py:189]: we are double resolving defaults int he process_config case right? seems like this belongs in the if block 

might make sense to make an entry point like `process_config` that captures the validate then resolve defaults pattern

File: python_modules/dagster/dagster/core/system_config/objects.py [hunk 3]
```diff
@@ -152,18 +167,28 @@ def build(
         run_config_schema = pipeline_def.get_run_config_schema(mode)
 
         if run_config_schema.config_mapping:
-            outer_evr = process_config(
-                run_config_schema.config_mapping.config_schema.config_type,
-                run_config,
-            )
+            if run_config_schema.config_mapping.enable_pre_processing:
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/system_config/objects.py:170]: if we are putting this on `ConfigMapping` I am guessing we are not respecting this in other places that config mappings are evaluated.

likely makes sense to add a method on to `ConfigMapping` that handles this pattern

**Review by alangenfeld — CHANGES_REQUESTED** [review]

File: python_modules/dagster/dagster/core/system_config/objects.py (hunk 3 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/core/system_config/objects.py:170]: Do you mean a method on ConfigMapping that resolves its own config? Like `ConfigMapping.resolve_from_outer` or something?

Comment (alangenfeld) [reply python_modules/dagster/dagster/core/system_config/objects.py]: yeah to perform the operation `ConfigMapping` exists to do, `ConfigMapping.map_config` though that name feels a little rough `ConfigMapping.map`, `ConfigMapping.process`, `ConfigMapping.resolve`, `ConfigMapping.execute`

**Comment by dpeng817** [comment]
I added a `ConfigMapping.resolve` method. Not sure how I feel about the bypass flag for the validation step, but it seems necessary to use from all paths.

File: python_modules/dagster/dagster/core/system_config/objects.py (hunk 3 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/core/system_config/objects.py]: Ended up going with `ConfigMapping.resolve`

File: python_modules/dagster/dagster/core/definitions/config.py [hunk 4]
```diff
@@ -35,15 +43,53 @@ class ConfigMapping(
...
+            outer_config = outer_evr.value
+            if not self.receive_processed_config_values:
+                outer_config = resolve_defaults(
+                    cast(ConfigType, self.config_schema.config_type),
+                    outer_config,
+                ).value
+        else:
+            outer_config = config
+
+        return self.config_fn(outer_config)
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/definitions/config.py:95]: I think just having two different methods would be cleaner than this bool forking most of one fn 

but also this else case seems really bizarre - given the names of the args its unclear why default resolution still doesn't happen

File: python_modules/dagster/dagster/core/definitions/graph.py [hunk 5]
```diff
@@ -799,8 +799,7 @@ def config_fn(x):
         )
 
     return ConfigMapping(
-        config_fn=config_fn,
-        config_schema=config_schema,
+        config_fn=config_fn, config_schema=config_schema, receive_processed_config_values=False
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/definitions/graph.py:802]: leave a comment?

File: python_modules/dagster/dagster/core/system_config/composite_descent.py [hunk 6]
```diff
@@ -176,8 +176,8 @@ def _get_mapped_solids_dict(
     with user_code_error_boundary(
         DagsterConfigMappingFunctionError, _get_error_lambda(current_stack)
     ):
-        mapped_solids_config = graph_def.config_mapping.config_fn(
-            config_mapped_solid_config.value.get("config", {})
+        mapped_solids_config = graph_def.config_mapping.resolve(
+            config_mapped_solid_config.value.get("config", {}), config_has_been_validated=True
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/system_config/composite_descent.py:180]: how do we handle enums in composite descent? is it under test at all?

Comment (alangenfeld) [reply python_modules/dagster/dagster/core/system_config/composite_descent.py:180]: If you don't  want to chase this down in this PR, just add a comment and keep calling config_fn directly

**Review by alangenfeld — APPROVED** [review]
drop the `config_has_been_validated` wonkiness and this is g2g

**Pull request merged**